packages/beacon-node/src/chain/blocks/blockInput/blockInput.ts

```diff
@@ -0,0 +1,786 @@
+      }
+    }
+
+    const hasAllData = props.block.message.body.blobKzgCommitments.length === 0 || this.state.hasAllData;
+
+    this.state = {
+      ...this.state,
+      hasBlock: true,
+      hasAllData,
+      block: props.block,
+      source: props.source,
+      timeCompleteSec: hasAllData ? props.source.seenTimestampSec : undefined,
+    } as BlockInputColumnsState;
+
+    this.blockPromise.resolve(props.block);
+  }
+
+  addColumn({blockRootHex, columnSidecar, source, seenTimestampSec, peerIdStr}: AddColumn): void {
+    if (blockRootHex !== this.blockRootHex) {
+      throw new BlockInputError(
+        {
+          code: BlockInputErrorCode.MISMATCHED_ROOT_HEX,
+          blockInputRoot: this.blockRootHex,
+          mismatchedRoot: blockRootHex,
+          source: source,
+          peerId: `${peerIdStr}`,
+        },
+        "Column BeaconBlockHeader blockRootHex does not match BlockInput.blockRootHex"
+      );
+    }
+
+    if (this.state.hasBlock) {
+      assertBlockAndColumnArePaired(this.blockRootHex, this.state.block, columnSidecar);
+    }
+
+    this.columnsCache.set(columnSidecar.index, {columnSidecar, source, seenTimestampSec, peerIdStr});
+
+    // check if we have freshly completed sampled or custody columns
+    // eg: sampledComplete == true && sampledColumns !== null
+
+    let hasAllData = this.state.hasAllData;
+    let sampledColumns: fulu.DataColumnSidecars | null = null;
+    // biome-ignore lint/suspicious/noConfusingLabels: <explanation>
+    maybeSampleComplete: if (!hasAllData) {
+      sampledColumns = [];
+      for (const index of this.custodyConfig.sampledColumns) {
+        const column = this.columnsCache.get(index);
+        if (column) {
+          sampledColumns.push(column.columnSidecar);
+        } else {
+          break maybeSampleComplete;
+        }
+      }
+      hasAllData = true;
+    }
+
+    this.state = {
+      ...this.state,
+      hasAllData: hasAllData || this.state.hasAllData,
+      timeCompleteSec: hasAllData ? seenTimestampSec : undefined,
+    } as BlockInputColumnsState;
+
+    if (hasAllData && sampledColumns !== null) {
+      this.dataPromise.resolve(sampledColumns);
+    }
+  }
+
+  hasColumn(columnIndex: number): boolean {
+    return this.columnsCache.has(columnIndex);
+  }
+
+  getCustodyColumns(): fulu.DataColumnSidecars {
+    const columns: fulu.DataColumnSidecars = [];
+    for (const index of this.custodyConfig.sampledColumns) {
+      const column = this.columnsCache.get(index);
+      if (column) {
+        columns.push(column.columnSidecar);
+      }
+    }
+    return columns;
+  }
+
+  getSampledColumns(): fulu.DataColumnSidecars {
+    const columns: fulu.DataColumnSidecars = [];
+    for (const index of this.custodyConfig.sampledColumns) {
+      const column = this.columnsCache.get(index);
+      if (column) {
+        columns.push(column.columnSidecar);
+      }
+    }
+    return columns;
+  }
```

Comment:
I have these optionally throwing so they can be used internally and externally (with surety that they are all there).

Reply:
Can DRY this with a higher-order implementation

```suggestion
  private makeColumnGetter(type: "sampled" | "custody"): () => fulu.DataColumnSidecars {
    return () => {
      const columns: fulu.DataColumnSidecars = [];
      for (const index of this.custodyConfig[`${type}Columns`]) {
        const column = this.columnsCache.get(index);
        if (column) {
          columns.push(column.columnSidecar);
        }
      }
      return columns;
    };
  }

  getCustodyColumns = this.makeColumnGetter("custody").bind(this);

  getSampledColumns = this.makeColumnGetter("sampled").bind(this);
```

Reply:
not a fan of throwing or higher order function here. its DRY yes but this kind of fanciness leads to worse compiled code and imo just worse readability considering its a simple map/filter. (consider that the custodyConfig property access is dynamic in the higher order impl)